token-metadata/js/package.json

```diff
@@ -0,0 +1,79 @@
+{
+    "name": "@solana/spl-token-metadata",
+    "description": "SPL Token Metadata Interface JS API",
+    "version": "0.0.1",
+    "author": "Solana Labs Maintainers <maintainers@solanalabs.com>",
+    "repository": "https://github.com/solana-labs/solana-program-library",
+    "license": "Apache-2.0",
+    "type": "module",
+    "sideEffects": false,
+    "engines": {
+        "node": ">=16"
+    },
+    "files": [
+        "lib",
+        "src",
+        "LICENSE",
+        "README.md"
+    ],
+    "publishConfig": {
+        "access": "public"
+    },
+    "main": "./lib/cjs/index.js",
+    "module": "./lib/esm/index.js",
+    "types": "./lib/types/index.d.ts",
+    "exports": {
+        "types": "./lib/types/index.d.ts",
+        "require": "./lib/cjs/index.js",
+        "import": "./lib/esm/index.js"
+    },
+    "scripts": {
+        "build": "tsc --build --verbose tsconfig.all.json",
+        "clean": "shx rm -rf lib **/*.tsbuildinfo || true",
+        "deploy": "npm run deploy:docs",
+        "deploy:docs": "npm run docs && gh-pages --dest token-metadata/js --dist docs --dotfiles",
+        "docs": "shx rm -rf docs && typedoc && shx cp .nojekyll docs/",
+        "fmt": "prettier --write '{*,**/*}.{ts,tsx,js,jsx,json}'",
+        "lint": "prettier --check '{*,**/*}.{ts,tsx,js,jsx,json}' && eslint --max-warnings 0 .",
+        "lint:fix": "npm run fmt && eslint --fix .",
+        "nuke": "shx rm -rf node_modules package-lock.json || true",
+        "postbuild": "shx echo '{ \"type\": \"commonjs\" }' > lib/cjs/package.json",
+        "reinstall": "npm run nuke && npm install",
+        "release": "npm run clean && npm run build",
+        "test": "mocha test",
+        "watch": "tsc --build --verbose --watch tsconfig.all.json"
+    },
+    "peerDependencies": {
+        "@solana/web3.js": "^1.47.4"
+    },
+    "dependencies": {
+        "@solana/spl-type-length-value": "link:../../libraries/type-length-value/js",
+        "borsh": "^1.0.0"
```

Comment:
Somewhere else in this repo, `https://www.npmjs.com/package/@coral-xyz/borsh` is being used. 

There wasn't any docs for coral-xyz package, and this packaged looked easier to use + more downloads on npm. 

Any rules surrounding this?

Reply:
We _do_ use buffer-layout in `@solana/spl-token` though

https://github.com/solana-labs/solana-program-library/blob/5ce6eace7c128c3eb5686e02de983ee431b73ce1/token/js/src/extensions/transferFee/instructions.ts#L39-L49

Reply:
I tried to switch to `@solana/buffer-layouts` but two main notes:

1. `string` support is a bit clunky. I have to manually calculate the buffer length. The string length is variable, and doesn't seem like the library can handle it and always expect a buffer first. Its possible to over-allocate and trim after to what was used, but seems less efficient. Also seem to have to manually add the string length as a u32 before the data. (I also tried the utf8 type in the buffer-layouts, but didn't solve the above problems and kept giving me typescripts errors)
```
    const layout = struct<RemoveKeyData>([bool('idempotent'), u32('fieldLength'), cstr('field')]);

    const fieldLength = Buffer.byteLength(field, 'utf8');

    const buffer = Buffer.alloc(1 + 4 + fieldLength);
```

2. There doesn't seem to be a way to encode optional parameters in `@solana/buffer-layouts` such as whats needed for the `Emit` instruction.

Reply:
Also I don't think the borsh library can encode the "Field" correctly, tried for a while but can't handle it. It can do enums, just not enums with a dynamic `Key(string)` value.

Not sure if you have encouraged similar problems before @buffalojoec ? 

I have 2 potential solutions:

1. Stick with `@solana/buffer-layouts` - but its going to be messy. Think i will have to manually handle the optional parameters and the `Field` parameter.

2. Try this alternative package - Seems to work similar to the existing `borsh` package, but we can define a custom encoder for the `Field` in `UpdateField`
https://www.npmjs.com/package/@dao-xyz/borsh


I'm leaning towards doing 1, avoid having to use another package and we still have to right custom serialiser/deserialiser - but thought I check before I head down this path.